NPM install fails with Node 4.0

make: Entering directory `/data/dev/zk/node_modules/zookeeper/build'
  ACTION binding_gyp_libzk_target_build_zk_client_lib .
Downloading zookeeper-3.4.3 from http://apache.mirrors.tds.net/zookeeper/zookeeper-3.4.3/zookeeper-3.4.3.tar.gz

gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Error is not recoverable: exiting now
Unable to build zookeeper library
make: *** [.] Error 1
make: Leaving directory `/data/dev/zk/node_modules/zookeeper/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/root/.nvm/versions/node/v4.0.0/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:270:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Linux 2.6.32-504.30.3.el6.x86_64
gyp ERR! command "/root/.nvm/versions/node/v4.0.0/bin/node" "/root/.nvm/versions/node/v4.0.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /data/dev/zk/node_modules/zookeeper
gyp ERR! node -v v4.0.0
gyp ERR! node-gyp -v v3.0.1
gyp ERR! not ok 
npm ERR! Linux 2.6.32-504.30.3.el6.x86_64
npm ERR! argv "/root/.nvm/versions/node/v4.0.0/bin/node" "/root/.nvm/versions/node/v4.0.0/bin/npm" "install" "zookeeper"
npm ERR! node v4.0.0
npm ERR! npm  v2.14.2
npm ERR! code ELIFECYCLE

npm ERR! zookeeper@3.4.6 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the zookeeper@3.4.6 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the zookeeper package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls zookeeper
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /data/dev/zk/npm-debug.log

--- a/package.json
+++ b/package.json
@@ -21,7 +21,7 @@
   ,"keywords": ["apache", "zookeeper", "client"]
   ,"dependencies": {
     "async": "~0.2.6"
-    ,"nan": "^1.8.4"
+    ,"nan": "~2.0.9"
     ,"underscore": "*"
   }
   ,"devDependencies": {

--- a/src/buffer_compat.h
+++ b/src/buffer_compat.h
@@ -6,47 +6,18 @@
 #include <node_version.h>
 #include <v8.h>
 
-#if NODE_MAJOR_VERSION < 11 && NODE_MINOR_VERSION < 3
-
-char *BufferData(node::Buffer *b) {
-    return b->data();
-}
-size_t BufferLength(node::Buffer *b) {
-    return b->length();
-}
-char *BufferData(v8::Local<v8::Object> buf_obj) {
-    v8::HandleScope scope;
-    node::Buffer *buf = node::ObjectWrap::Unwrap<node::Buffer>(buf_obj);
-    return buf->data();
-}
-size_t BufferLength(v8::Local<v8::Object> buf_obj) {
-    v8::HandleScope scope;
-    node::Buffer *buf = node::ObjectWrap::Unwrap<node::Buffer>(buf_obj);
-    return buf->length();
-}
-
-Local<Object> BufferNew(const char* bytes, int length) {
-    Buffer* b = Buffer::New(value_len);
-    memcpy(BufferData(b), value, value_len);
-    return Local<Object>::New(b->handle_);
-}
-
-#else // NODE_VERSION
 #include "nan.h"
 
-Local<Object> BufferNew(const char* bytes, int length) {
-    return NanNewBufferHandle(bytes, length);
+Nan::MaybeLocal<v8::Object> BufferNew(const char* bytes, uint32_t length) {
+    return Nan::CopyBuffer(bytes, length);
 }
 
 char *BufferData(v8::Local<v8::Object> buf_obj) {
-    NanScope();
     return node::Buffer::Data(buf_obj);
 }
+
 size_t BufferLength(v8::Local<v8::Object> buf_obj) {
-    NanScope();
     return node::Buffer::Length(buf_obj);
 }
 
-#endif // NODE_VERSION
-
-#endif//BUFFER_COMPAT_H
+#endif // BUFFER_COMPAT_H

--- a/src/node-zk.cpp
+++ b/src/node-zk.cpp
@@ -319,6 +319,14 @@ public:
     }
 
     inline bool realInit (const char* hostPort, int session_timeout, clientid_t *client_id) {
+        bool need_timer_init = true;
+        if (zhandle) {
+            // In case this is not the first call to realInit,
+            // stop the current timer and skip re-initializing the timer
+            need_timer_init = true;
+            uv_timer_stop(&zk_timer);
+        }
+      
         myid = *client_id;
         zhandle = zookeeper_init(hostPort, main_watcher, session_timeout, &myid, this, 0);
         if (!zhandle) {
@@ -327,8 +335,10 @@ public:
         }
         Ref();
 
-        uv_timer_init(uv_default_loop(), &zk_timer);
-        zk_io.data = zk_timer.data = this;
+        if (need_timer_init) {
+            uv_timer_init(uv_default_loop(), &zk_timer);
+            zk_io.data = zk_timer.data = this;
+        }
 
         yield();
         return true;

--- a/src/node-zk.cpp
+++ b/src/node-zk.cpp
@@ -46,17 +46,23 @@ namespace zk {
 #define _LLP_CAST_ (long long *)
 
 #define THROW_IF_NOT(condition, text) if (!(condition)) { \
-      return NanThrowError(text); \
+      return Nan::ThrowError(text); \
     }
 
 #define THROW_IF_NOT_R(condition, text) if (!(condition)) { \
-      NanThrowError(text); \
+      Nan::ThrowError(text); \
       return; \
     }
-
-#define DECLARE_STRING(ev) static Persistent<String> ev; 
-#define INITIALIZE_STRING(ev, str) NanAssignPersistent(ev, NanNew<String>(str)); 
-
+  
+  
+#define RETURN_THIS(info) info.GetReturnValue().Set(info.This())
+#define RETURN_VALUE(info, value) info.GetReturnValue().Set(value)
+  
+#define LOCAL_STRING(str) Nan::New<String>(str).ToLocalChecked()
+  
+#define DECLARE_STRING(ev) static Nan::Persistent<String> ev; 
+#define INITIALIZE_STRING(ev, str) ev.Reset(LOCAL_STRING(str)); 
+  
 DECLARE_STRING (on_closed);
 DECLARE_STRING (on_connected);
 DECLARE_STRING (on_connecting);
@@ -75,69 +81,69 @@ DECLARE_SYMBOL (HIDDEN_PROP_HANDBACK);
 #define ZOOKEEPER_PASSWORD_BYTE_COUNT 16
 
 struct completion_data {
-    NanCallback *cb;
+    Nan::Callback *cb;
     int32_t type;
     void *data;
 };
 
-class ZooKeeper: public ObjectWrap {
+class ZooKeeper: public Nan::ObjectWrap {
 public:
     static void Initialize (v8::Handle<v8::Object> target) {
-        NanScope();
+        Nan::HandleScope scope;
 
-        Local<FunctionTemplate> constructor_template = NanNew<FunctionTemplate>(New);
-        constructor_template->SetClassName(NanNew("ZooKeeper"));
+        Local<FunctionTemplate> constructor_template = Nan::New<FunctionTemplate>(New);
+        constructor_template->SetClassName(LOCAL_STRING("ZooKeeper"));
         constructor_template->InstanceTemplate()->SetInternalFieldCount(1);
 
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "init", Init);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "close", Close);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "a_create", ACreate);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "a_exists", AExists);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "aw_exists", AWExists);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "a_get", AGet);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "aw_get", AWGet);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "a_get_children", AGetChildren);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "aw_get_children", AWGetChildren);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "a_get_children2", AGetChildren2);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "aw_get_children2", AWGetChildren2);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "a_set", ASet);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "a_delete_", ADelete);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "s_delete_", Delete);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "a_get_acl", AGetAcl);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "a_set_acl", ASetAcl);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "add_auth", AddAuth);
-        NODE_SET_PROTOTYPE_METHOD(constructor_template, "a_sync", ASync);
+        Nan::SetPrototypeMethod(constructor_template,  "init",  Init);
+        Nan::SetPrototypeMethod(constructor_template,  "close",  Close);
+        Nan::SetPrototypeMethod(constructor_template,  "a_create",  ACreate);
+        Nan::SetPrototypeMethod(constructor_template,  "a_exists",  AExists);
+        Nan::SetPrototypeMethod(constructor_template,  "aw_exists",  AWExists);
+        Nan::SetPrototypeMethod(constructor_template,  "a_get",  AGet);
+        Nan::SetPrototypeMethod(constructor_template,  "aw_get",  AWGet);
+        Nan::SetPrototypeMethod(constructor_template,  "a_get_children",  AGetChildren);
+        Nan::SetPrototypeMethod(constructor_template,  "aw_get_children",  AWGetChildren);
+        Nan::SetPrototypeMethod(constructor_template,  "a_get_children2",  AGetChildren2);
+        Nan::SetPrototypeMethod(constructor_template,  "aw_get_children2",  AWGetChildren2);
+        Nan::SetPrototypeMethod(constructor_template,  "a_set",  ASet);
+        Nan::SetPrototypeMethod(constructor_template,  "a_delete_",  ADelete);
+        Nan::SetPrototypeMethod(constructor_template,  "s_delete_",  Delete);
+        Nan::SetPrototypeMethod(constructor_template,  "a_get_acl",  AGetAcl);
+        Nan::SetPrototypeMethod(constructor_template,  "a_set_acl",  ASetAcl);
+        Nan::SetPrototypeMethod(constructor_template,  "add_auth",  AddAuth);
+        Nan::SetPrototypeMethod(constructor_template,  "a_sync",  ASync);
 
         //what's the advantage of using constructor_template->PrototypeTemplate()->SetAccessor ?
-        constructor_template->InstanceTemplate()->SetAccessor(NanNew<String>("state"), StatePropertyGetter, 0, Local<Value>(), PROHIBITS_OVERWRITING, ReadOnly);
-        constructor_template->InstanceTemplate()->SetAccessor(NanNew<String>("client_id"), ClientidPropertyGetter, 0, Local<Value>(), PROHIBITS_OVERWRITING, ReadOnly);
-        constructor_template->InstanceTemplate()->SetAccessor(NanNew<String>("client_password"), ClientPasswordPropertyGetter, 0, Local<Value>(), PROHIBITS_OVERWRITING, ReadOnly);
-        constructor_template->InstanceTemplate()->SetAccessor(NanNew<String>("timeout"), SessionTimeoutPropertyGetter, 0, Local<Value>(), PROHIBITS_OVERWRITING, ReadOnly);
-        constructor_template->InstanceTemplate()->SetAccessor(NanNew<String>("is_unrecoverable"), IsUnrecoverablePropertyGetter, 0, Local<Value>(), PROHIBITS_OVERWRITING, ReadOnly);
+        Nan::SetAccessor(constructor_template->InstanceTemplate(), LOCAL_STRING("state"), StatePropertyGetter, 0, Local<Value>(), PROHIBITS_OVERWRITING, ReadOnly);
+        Nan::SetAccessor(constructor_template->InstanceTemplate(), LOCAL_STRING("client_id"), ClientidPropertyGetter, 0, Local<Value>(), PROHIBITS_OVERWRITING, ReadOnly);
+        Nan::SetAccessor(constructor_template->InstanceTemplate(), LOCAL_STRING("client_password"), ClientPasswordPropertyGetter, 0, Local<Value>(), PROHIBITS_OVERWRITING, ReadOnly);
+        Nan::SetAccessor(constructor_template->InstanceTemplate(), LOCAL_STRING("timeout"), SessionTimeoutPropertyGetter, 0, Local<Value>(), PROHIBITS_OVERWRITING, ReadOnly);
+        Nan::SetAccessor(constructor_template->InstanceTemplate(), LOCAL_STRING("is_unrecoverable"), IsUnrecoverablePropertyGetter, 0, Local<Value>(), PROHIBITS_OVERWRITING, ReadOnly);
 
 
         Local<Function> constructor = constructor_template->GetFunction();
 
         //extern ZOOAPI struct ACL_vector ZOO_OPEN_ACL_UNSAFE;
-        Local<Object> acl_open = NanNew<Object>();
-        acl_open->ForceSet(NanNew("perms"), NanNew<Integer>(ZOO_PERM_ALL), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
-        acl_open->ForceSet(NanNew("scheme"), NanNew("world"), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
-        acl_open->ForceSet(NanNew("auth"), NanNew("anyone"), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
-        constructor->ForceSet(NanNew("ZOO_OPEN_ACL_UNSAFE"), acl_open, static_cast<PropertyAttribute>(ReadOnly | DontDelete));
+        Local<Object> acl_open = Nan::New<Object>();
+        Nan::ForceSet(acl_open, LOCAL_STRING("perms"), Nan::New<Integer>(ZOO_PERM_ALL), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
+        Nan::ForceSet(acl_open, LOCAL_STRING("scheme"), LOCAL_STRING("world"), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
+        Nan::ForceSet(acl_open, LOCAL_STRING("auth"), LOCAL_STRING("anyone"), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
+        Nan::ForceSet(constructor, LOCAL_STRING("ZOO_OPEN_ACL_UNSAFE"), acl_open, static_cast<PropertyAttribute>(ReadOnly | DontDelete));
 
         //extern ZOOAPI struct ACL_vector ZOO_READ_ACL_UNSAFE;
-        Local<Object> acl_read = NanNew<Object>();
-        acl_read->ForceSet(NanNew<String>("perms"), NanNew<Integer>(ZOO_PERM_READ), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
-        acl_read->ForceSet(NanNew<String>("scheme"), NanNew<String>("world"), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
-        acl_read->ForceSet(NanNew<String>("auth"), NanNew<String>("anyone"), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
-        constructor->ForceSet(NanNew<String>("ZOO_READ_ACL_UNSAFE"), acl_read, static_cast<PropertyAttribute>(ReadOnly | DontDelete));
+        Local<Object> acl_read = Nan::New<Object>();
+        Nan::ForceSet(acl_read, LOCAL_STRING("perms"), Nan::New<Integer>(ZOO_PERM_READ), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
+        Nan::ForceSet(acl_read, LOCAL_STRING("scheme"), LOCAL_STRING("world"), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
+        Nan::ForceSet(acl_read, LOCAL_STRING("auth"), LOCAL_STRING("anyone"), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
+        Nan::ForceSet(constructor, LOCAL_STRING("ZOO_READ_ACL_UNSAFE"), acl_read, static_cast<PropertyAttribute>(ReadOnly | DontDelete));
 
         //extern ZOOAPI struct ACL_vector ZOO_CREATOR_ALL_ACL;
-        Local<Object> acl_creator = NanNew<Object>();
-        acl_creator->ForceSet(NanNew<String>("perms"), NanNew<Integer>(ZOO_PERM_ALL), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
-        acl_creator->ForceSet(NanNew<String>("scheme"), NanNew<String>("auth"), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
-        acl_creator->ForceSet(NanNew<String>("auth"), NanNew<String>(""), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
-        constructor->ForceSet(NanNew<String>("ZOO_CREATOR_ALL_ACL"), acl_creator, static_cast<PropertyAttribute>(ReadOnly | DontDelete));
+        Local<Object> acl_creator = Nan::New<Object>();
+        Nan::ForceSet(acl_creator, LOCAL_STRING("perms"), Nan::New<Integer>(ZOO_PERM_ALL), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
+        Nan::ForceSet(acl_creator, LOCAL_STRING("scheme"), LOCAL_STRING("auth"), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
+        Nan::ForceSet(acl_creator, LOCAL_STRING("auth"), LOCAL_STRING(""), static_cast<PropertyAttribute>(ReadOnly | DontDelete));
+        Nan::ForceSet(constructor, LOCAL_STRING("ZOO_CREATOR_ALL_ACL"), acl_creator, static_cast<PropertyAttribute>(ReadOnly | DontDelete));
 
 
         NODE_DEFINE_CONSTANT(constructor, ZOO_CREATED_EVENT);
@@ -222,16 +228,15 @@ public:
         NODE_DEFINE_CONSTANT(constructor, ZSESSIONMOVED);
 
 
-        target->Set(NanNew<String>("ZooKeeper"), constructor);
+        target->Set(LOCAL_STRING("ZooKeeper"), constructor);
     }
 
-    static NAN_METHOD(New) {
-        NanScope();
+    static void New(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         ZooKeeper *zk = new ZooKeeper();
 
-        zk->Wrap(args.This());
+        zk->Wrap(info.This());
         //zk->handle_.ClearWeak();
-        NanReturnThis();
+        RETURN_THIS(info);
     }
 
     void yield () {
@@ -344,62 +349,60 @@ public:
         return true;
     }
 
-    static NAN_METHOD(Init) {
-        NanScope();
+    static void Init(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+        THROW_IF_NOT(info.Length() >= 1, "Must pass ZK init object");
+        THROW_IF_NOT(info[0]->IsObject(), "Init argument must be an object");
+        Local<Object> arg = info[0]->ToObject();
 
-        THROW_IF_NOT(args.Length() >= 1, "Must pass ZK init object");
-        THROW_IF_NOT(args[0]->IsObject(), "Init argument must be an object");
-        Local<Object> arg = args[0]->ToObject();
-
-        int32_t debug_level = arg->Get(NanNew<String>("debug_level"))->ToInt32()->Value();
+        int32_t debug_level = arg->Get(LOCAL_STRING("debug_level"))->ToInt32()->Value();
         zoo_set_debug_level(static_cast<ZooLogLevel>(debug_level));
 
-        bool order = arg->Get(NanNew<String>("host_order_deterministic"))->ToBoolean()->BooleanValue();
+        bool order = arg->Get(LOCAL_STRING("host_order_deterministic"))->ToBoolean()->BooleanValue();
         zoo_deterministic_conn_order(order); // enable deterministic order
 
-        NanAsciiString _hostPort (arg->Get(NanNew<String>("connect"))->ToString());
-        int32_t session_timeout = arg->Get(NanNew<String>("timeout"))->ToInt32()->Value();
+        Nan::Utf8String _hostPort (arg->Get(LOCAL_STRING("connect"))->ToString());
+        int32_t session_timeout = arg->Get(LOCAL_STRING("timeout"))->ToInt32()->Value();
         if (session_timeout == 0) {
             session_timeout = 20000;
         }
 
         clientid_t local_client;
         ZERO_MEM (local_client);
-        v8::Local<v8::Value> v8v_client_id = arg->Get(NanNew<String>("client_id"));
-        v8::Local<v8::Value> v8v_client_password = arg->Get(NanNew<String>("client_password"));
+        v8::Local<v8::Value> v8v_client_id = arg->Get(LOCAL_STRING("client_id"));
+        v8::Local<v8::Value> v8v_client_password = arg->Get(LOCAL_STRING("client_password"));
         bool id_and_password_defined = (!v8v_client_id->IsUndefined() && !v8v_client_password->IsUndefined());
         bool id_and_password_undefined = (v8v_client_id->IsUndefined() && v8v_client_password->IsUndefined());
         THROW_IF_NOT ((id_and_password_defined || id_and_password_undefined), 
             "ZK init: client id and password must either be both specified or unspecified");
         if (id_and_password_defined) {
-          NanAsciiString password_check(v8v_client_password->ToString());
-          THROW_IF_NOT (password_check.length() == 2 * ZOOKEEPER_PASSWORD_BYTE_COUNT, 
-              "ZK init: password does not have correct length");
-          HexStringToPassword(v8v_client_password, local_client.passwd);
-          StringToId(v8v_client_id, &local_client.client_id);
+            Nan::Utf8String password_check(v8v_client_password->ToString());
+            THROW_IF_NOT (password_check.length() == 2 * ZOOKEEPER_PASSWORD_BYTE_COUNT, 
+                          "ZK init: password does not have correct length");
+            HexStringToPassword(v8v_client_password, local_client.passwd);
+            StringToId(v8v_client_id, &local_client.client_id);
         }
 
-        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(args.This());
+        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(info.This());
         assert(zk);
 
         if (!zk->realInit(*_hostPort, session_timeout, &local_client)) {
-            NanReturnValue(ErrnoException(errno, "zookeeper_init", "failed to init", __FILE__));
+            RETURN_VALUE(info, Nan::NanErrnoException(errno, "zookeeper_init", "failed to init", __FILE__));
         } else {
-            NanReturnThis();
+            RETURN_THIS(info);
         }
     }
 
     static void main_watcher (zhandle_t *zzh, int type, int state, const char *path, void* context) {
+        Nan::HandleScope scope;
         LOG_DEBUG(("main watcher event: type=%d, state=%d, path=%s", type, state, (path ? path: "null")));
         ZooKeeper *zk = static_cast<ZooKeeper *>(context);
-        NanScope();
 
         if (type == ZOO_SESSION_EVENT) {
             if (state == ZOO_CONNECTED_STATE) {
                 zk->myid = *(zoo_client_id(zzh));
-                zk->DoEmitPath(NanNew(on_connected), path);
+                zk->DoEmitPath(Nan::New(on_connected), path);
             } else if (state == ZOO_CONNECTING_STATE) {
-                zk->DoEmitPath (NanNew(on_connecting), path);
+                zk->DoEmitPath (Nan::New(on_connecting), path);
             } else if (state == ZOO_AUTH_FAILED_STATE) {
                 LOG_ERROR(("Authentication failure. Shutting down...\n"));
                 zk->realClose(ZOO_AUTH_FAILED_STATE);
@@ -408,99 +411,99 @@ public:
                 zk->realClose(ZOO_EXPIRED_SESSION_STATE);
             }
         } else if (type == ZOO_CREATED_EVENT) {
-            zk->DoEmitPath(NanNew(on_event_created), path);
+            zk->DoEmitPath(Nan::New(on_event_created), path);
         } else if (type == ZOO_DELETED_EVENT) {
-            zk->DoEmitPath(NanNew(on_event_deleted), path);
+            zk->DoEmitPath(Nan::New(on_event_deleted), path);
         } else if (type == ZOO_CHANGED_EVENT) {
-            zk->DoEmitPath(NanNew(on_event_changed), path);
+            zk->DoEmitPath(Nan::New(on_event_changed), path);
         } else if (type == ZOO_CHILD_EVENT) {
-            zk->DoEmitPath(NanNew(on_event_child), path);
+            zk->DoEmitPath(Nan::New(on_event_child), path);
         } else if (type == ZOO_NOTWATCHING_EVENT) {
-            zk->DoEmitPath(NanNew(on_event_notwatching), path);
+            zk->DoEmitPath(Nan::New(on_event_notwatching), path);
         } else {
             LOG_WARN(("Unknonwn watcher event type %s",type));
         }
     }
 
     static Local<String> idAsString (int64_t id) {
-        NanEscapableScope();
+        Nan::EscapableHandleScope scope;
         char idbuff [128] = {0};
         sprintf(idbuff, "%llx", _LL_CAST_ id);
-        return NanEscapeScope(NanNew<String>(idbuff));
+        return scope.Escape(LOCAL_STRING(idbuff));
     }
 
     static void StringToId (v8::Local<v8::Value> s, int64_t *id) {
-        NanAsciiString a(s->ToString());
+        Nan::Utf8String a(s->ToString());
         sscanf(*a, "%llx", _LLP_CAST_ id);
     }
 
     static Local<String> PasswordToHexString (const char *p) {
-        NanEscapableScope();
+        Nan::EscapableHandleScope scope;
         char buff[ZOOKEEPER_PASSWORD_BYTE_COUNT * 2 + 1], *b = buff;
         for (int i = 0; i < ZOOKEEPER_PASSWORD_BYTE_COUNT; ++i) {
             ucharToHex((unsigned char *) (p + i), b);
             b += 2;
         }
         buff[ZOOKEEPER_PASSWORD_BYTE_COUNT * 2] = '\0';
-        return NanEscapeScope(NanNew<String>(buff));
+        return scope.Escape(LOCAL_STRING(buff));
     }
 
     static void HexStringToPassword (v8::Local<v8::Value> s, char *p) {
-        NanAsciiString a(s->ToString());
+        Nan::Utf8String a(s->ToString());
         char *hex = *a;
         for (int i = 0; i < ZOOKEEPER_PASSWORD_BYTE_COUNT; ++i) {
-          hexToUchar(hex, (unsigned char *)p+i);
-          hex += 2;
+            hexToUchar(hex, (unsigned char *)p+i);
+            hex += 2;
         }
     }
 
     void DoEmitPath (Local<String> event_name, const char* path = NULL) {
-        NanScope();
+        Nan::HandleScope scope;
         Local<Value> str;
 
         if (path != 0) {
-            str = NanNew<String>(path);
-            LOG_DEBUG(("calling Emit(%s, path='%s')", *NanUtf8String(NanNew(event_name)), path));
+            str = LOCAL_STRING(path);
+            LOG_DEBUG(("calling Emit(%s, path='%s')", *Nan::Utf8String(event_name), path));
         } else {
-            str = NanUndefined();
-            LOG_DEBUG(("calling Emit(%s, path=null)", *NanUtf8String(NanNew(event_name))));
+            str = Nan::Undefined();
+            LOG_DEBUG(("calling Emit(%s, path=null)", *Nan::Utf8String(event_name)));
         }
 
         this->DoEmit(event_name, str);
     }
 
     void DoEmitClose (Local<String> event_name, int code) {
-        NanScope();
-        Local<Value> v8code = NanNew<Number>(code);
+        Nan::HandleScope scope;
+        Local<Value> v8code = Nan::New<Number>(code);
 
         this->DoEmit(event_name, v8code);
     }
 
-    void DoEmit (Local<String> event_name, Handle<Value> data) {
-        NanScope();
-        Local<Object> thisObj = NanObjectWrapHandle(this);
+    void DoEmit (Local<String> event_name, Local<Value> data) {
+        Nan::HandleScope scope;
+        Local<Object> thisObj = this->handle();
 
         Local<Value> argv[3];
         argv[0] = event_name;
         argv[1] = thisObj;
-        argv[2] = NanNew<Value>(data);
+        argv[2] = data;
 
-        NanMakeCallback(thisObj, "emit", 3, argv);
+        Nan::MakeCallback(thisObj, "emit", 3, argv);
     }
 
-#define CALLBACK_PROLOG(args) \
-        NanScope(); \
-        NanCallback *callback = (NanCallback*)(cb); \
+#define CALLBACK_PROLOG(info) \
+        Nan::HandleScope scope; \
+        Nan::Callback *callback = (Nan::Callback*)(cb); \
         assert (callback); \
-        Local<Value> lv = callback->GetFunction()->GetHiddenValue(NanNew(HIDDEN_PROP_ZK)); \
+        Local<Value> lv = callback->GetFunction()->GetHiddenValue(Nan::New(HIDDEN_PROP_ZK)); \
         /*(*callback)->DeleteHiddenValue(HIDDEN_PROP_ZK);*/ \
         Local<Object> zk_handle = Local<Object>::Cast(lv); \
         ZooKeeper *zkk = ObjectWrap::Unwrap<ZooKeeper>(zk_handle); \
         assert(zkk);\
-        assert(NanObjectWrapHandle(zkk) == zk_handle);  \
-        Local<Value> argv[args]; \
-        argv[0] = NanNew<Int32>(rc);           \
-        argv[1] = NanNew<String>(zerror(rc))
+        assert(zkk->handle() == zk_handle);        \
+        Local<Value> argv[info]; \
+        argv[0] = Nan::New<Int32>(rc);           \
+        argv[1] = LOCAL_STRING(zerror(rc))
 
 #define CALLBACK_EPILOG() \
         callback->Call(sizeof(argv)/sizeof(argv[0]), argv); \
@@ -510,55 +513,53 @@ public:
         callback->Call(sizeof(argv)/sizeof(argv[0]), argv); \
 
 #define A_METHOD_PROLOG(nargs) \
-        NanScope();                                                       \
-        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(args.This()); \
+        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(info.This()); \
         assert(zk);\
-        THROW_IF_NOT (args.Length() >= nargs, "expected "#nargs" arguments") \
-        assert (args[nargs-1]->IsFunction()); \
-        NanCallback *cb = new NanCallback(args[nargs-1].As<Function>()); \
-        cb->GetFunction()->SetHiddenValue(NanNew(HIDDEN_PROP_ZK), NanObjectWrapHandle(zk)); \
+        THROW_IF_NOT (info.Length() >= nargs, "expected "#nargs" arguments") \
+        assert (info[nargs-1]->IsFunction()); \
+        Nan::Callback *cb = new Nan::Callback(info[nargs-1].As<Function>()); \
+        cb->GetFunction()->SetHiddenValue(Nan::New(HIDDEN_PROP_ZK), zk->handle()); \
 
 #define METHOD_EPILOG(call) \
         int ret = (call); \
-        NanReturnValue(NanNew<Int32>(ret))
+        RETURN_VALUE(info, Nan::New<Int32>(ret))
 
-#define WATCHER_PROLOG(args) \
+#define WATCHER_PROLOG(info) \
         if (zoo_state(zh) == ZOO_EXPIRED_SESSION_STATE) { return; } \
-        NanScope();                                                    \
-        NanCallback *callback = (NanCallback*)(watcherCtx);            \
+        Nan::HandleScope scope;                                                    \
+        Nan::Callback *callback = (Nan::Callback*)(watcherCtx);            \
         assert (callback); \
-        Local<Value> lv_zk = callback->GetFunction()->GetHiddenValue(NanNew(HIDDEN_PROP_ZK)); \
+        Local<Value> lv_zk = callback->GetFunction()->GetHiddenValue(Nan::New(HIDDEN_PROP_ZK)); \
         /* (*callback)->DeleteHiddenValue(HIDDEN_PROP_ZK); */ \
         Local<Object> zk_handle = Local<Object>::Cast(lv_zk); \
         ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(zk_handle); \
         assert(zk);\
-        assert(NanObjectWrapHandle(zk) == zk_handle);   \
+        assert(zk->handle() == zk_handle); \
         assert(zk->zhandle == zh); \
-        Local<Value> argv[args]; \
-        argv[0] = NanNew<Integer>(type);   \
-        argv[1] = NanNew<Integer>(state);  \
-        argv[2] = NanNew<String>(path);                                 \
-        Local<Value> lv_hb = callback->GetFunction()->GetHiddenValue(NanNew(HIDDEN_PROP_HANDBACK)); \
+        Local<Value> argv[info]; \
+        argv[0] = Nan::New<Integer>(type);   \
+        argv[1] = Nan::New<Integer>(state);  \
+        argv[2] = LOCAL_STRING(path);                                 \
+        Local<Value> lv_hb = callback->GetFunction()->GetHiddenValue(Nan::New(HIDDEN_PROP_HANDBACK)); \
         /* (*callback)->DeleteHiddenValue(HIDDEN_PROP_HANDBACK); */ \
-        argv[3] = NanUndefined();    \
+        argv[3] = Nan::Undefined();    \
         if (!lv_hb.IsEmpty()) argv[3] = lv_hb
 
 #define AW_METHOD_PROLOG(nargs) \
-        NanScope();                                                       \
-        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(args.This()); \
+        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(info.This()); \
         assert(zk);\
-        THROW_IF_NOT (args.Length() >= nargs, "expected at least "#nargs" arguments") \
-        assert (args[nargs-1]->IsFunction()); \
-        NanCallback *cb = new NanCallback (args[nargs-1].As<Function>()); \
-        cb->GetFunction()->SetHiddenValue(NanNew(HIDDEN_PROP_ZK), NanObjectWrapHandle(zk)); \
+        THROW_IF_NOT (info.Length() >= nargs, "expected at least "#nargs" arguments") \
+        assert (info[nargs-1]->IsFunction()); \
+        Nan::Callback *cb = new Nan::Callback (info[nargs-1].As<Function>()); \
+        cb->GetFunction()->SetHiddenValue(Nan::New(HIDDEN_PROP_ZK), zk->handle()); \
         \
-        assert (args[nargs-2]->IsFunction()); \
-        NanCallback *cbw = new NanCallback (args[nargs-2].As<Function>()); \
-        cbw->GetFunction()->SetHiddenValue(NanNew(HIDDEN_PROP_ZK), NanObjectWrapHandle(zk))
+        assert (info[nargs-2]->IsFunction()); \
+        Nan::Callback *cbw = new Nan::Callback (info[nargs-2].As<Function>()); \
+        cbw->GetFunction()->SetHiddenValue(Nan::New(HIDDEN_PROP_ZK), zk->handle())
 
 /*
-        if (args.Length() > nargs) { \
-            (*cbw)->SetHiddenValue(HIDDEN_PROP_HANDBACK, args[nargs]); \
+        if (info.Length() > nargs) { \
+            (*cbw)->SetHiddenValue(HIDDEN_PROP_HANDBACK, info[nargs]); \
         }
 */
 
@@ -570,21 +571,21 @@ public:
         LOG_DEBUG(("rc=%d, rc_string=%s, path=%s, data=%lp", rc, zerror(rc), value, cb));
 
         CALLBACK_PROLOG(3);
-        argv[2] = NanNew<String>(value);
+        argv[2] = LOCAL_STRING(value);
         CALLBACK_EPILOG();
     }
 
-    static NAN_METHOD(ACreate) {
+    static void ACreate(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         A_METHOD_PROLOG(4);
 
-        NanUtf8String _path (args[0]->ToString());
-        uint32_t flags = args[2]->ToUint32()->Uint32Value();
+        Nan::Utf8String _path (info[0]->ToString());
+        uint32_t flags = info[2]->ToUint32()->Uint32Value();
 
-        if (Buffer::HasInstance(args[1])) { // buffer
-            Local<Object> _data = args[1]->ToObject();
+        if (Buffer::HasInstance(info[1])) { // buffer
+            Local<Object> _data = info[1]->ToObject();
             METHOD_EPILOG(zoo_acreate(zk->zhandle, *_path, BufferData(_data), BufferLength(_data), &ZOO_OPEN_ACL_UNSAFE, flags, string_completion, cb));
         } else {    // other
-            NanUtf8String _data (args[1]->ToString());
+            Nan::Utf8String _data (info[1]->ToString());
             METHOD_EPILOG(zoo_acreate(zk->zhandle, *_path, *_data, _data.length(), &ZOO_OPEN_ACL_UNSAFE, flags, string_completion, cb));
         }
     }
@@ -605,10 +606,10 @@ public:
         free(d);
     }
 
-    static NAN_METHOD(ADelete) {
+    static void ADelete(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         A_METHOD_PROLOG(3);
-        NanUtf8String _path (args[0]->ToString());
-        uint32_t version = args[1]->ToUint32()->Uint32Value();
+        Nan::Utf8String _path (info[0]->ToString());
+        uint32_t version = info[1]->ToUint32()->Uint32Value();
 
         struct completion_data *data = (struct completion_data *) malloc(sizeof(struct completion_data));
         data->cb = cb;
@@ -619,44 +620,44 @@ public:
     }
 
     Local<Object> createStatObject (const struct Stat *stat) {
-        NanEscapableScope();
-        Local<Object> o = NanNew<Object>();
-        o->ForceSet(NanNew<String>("czxid"), NanNew<Number>(stat->czxid), ReadOnly);
-        o->ForceSet(NanNew<String>("mzxid"), NanNew<Number>(stat->mzxid), ReadOnly);
-        o->ForceSet(NanNew<String>("pzxid"), NanNew<Number>(stat->pzxid), ReadOnly);
-        o->ForceSet(NanNew<String>("dataLength"), NanNew<Integer>(stat->dataLength), ReadOnly);
-        o->ForceSet(NanNew<String>("numChildren"), NanNew<Integer>(stat->numChildren), ReadOnly);
-        o->ForceSet(NanNew<String>("version"), NanNew<Integer>(stat->version), ReadOnly);
-        o->ForceSet(NanNew<String>("cversion"), NanNew<Integer>(stat->cversion), ReadOnly);
-        o->ForceSet(NanNew<String>("aversion"), NanNew<Integer>(stat->aversion), ReadOnly);
-        o->ForceSet(NanNew<String>("ctime"), NODE_UNIXTIME_V8(stat->ctime/1000.), ReadOnly);
-        o->ForceSet(NanNew<String>("mtime"), NODE_UNIXTIME_V8(stat->mtime/1000.), ReadOnly);
-        o->ForceSet(NanNew<String>("ephemeralOwner"), idAsString(stat->ephemeralOwner), ReadOnly);
-        o->ForceSet(NanNew<String>("createdInThisSession"), NanNew<Boolean>(myid.client_id == stat->ephemeralOwner), ReadOnly);
-        return NanEscapeScope(o);
+        Nan::EscapableHandleScope scope;
+        Local<Object> o = Nan::New<Object>();
+        Nan::ForceSet(o, LOCAL_STRING("czxid"), Nan::New<Number>(stat->czxid), ReadOnly);
+        Nan::ForceSet(o, LOCAL_STRING("mzxid"), Nan::New<Number>(stat->mzxid), ReadOnly);
+        Nan::ForceSet(o, LOCAL_STRING("pzxid"), Nan::New<Number>(stat->pzxid), ReadOnly);
+        Nan::ForceSet(o, LOCAL_STRING("dataLength"), Nan::New<Integer>(stat->dataLength), ReadOnly);
+        Nan::ForceSet(o, LOCAL_STRING("numChildren"), Nan::New<Integer>(stat->numChildren), ReadOnly);
+        Nan::ForceSet(o, LOCAL_STRING("version"), Nan::New<Integer>(stat->version), ReadOnly);
+        Nan::ForceSet(o, LOCAL_STRING("cversion"), Nan::New<Integer>(stat->cversion), ReadOnly);
+        Nan::ForceSet(o, LOCAL_STRING("aversion"), Nan::New<Integer>(stat->aversion), ReadOnly);
+        Nan::ForceSet(o, LOCAL_STRING("ctime"), NODE_UNIXTIME_V8(stat->ctime/1000.), ReadOnly);
+        Nan::ForceSet(o, LOCAL_STRING("mtime"), NODE_UNIXTIME_V8(stat->mtime/1000.), ReadOnly);
+        Nan::ForceSet(o, LOCAL_STRING("ephemeralOwner"), idAsString(stat->ephemeralOwner), ReadOnly);
+        Nan::ForceSet(o, LOCAL_STRING("createdInThisSession"), Nan::New<Boolean>(myid.client_id == stat->ephemeralOwner), ReadOnly);
+        return scope.Escape(o);
     }
 
     static void stat_completion (int rc, const struct Stat *stat, const void *cb) {
         CALLBACK_PROLOG(3);
 
         LOG_DEBUG(("rc=%d, rc_string=%s", rc, zerror(rc)));
-        argv[2] = rc == ZOK ? zkk->createStatObject (stat) : NanNull().As<Object>();
+        argv[2] = rc == ZOK ? zkk->createStatObject (stat) : Nan::Null().As<Object>();
 
         CALLBACK_EPILOG();
     }
 
-    static NAN_METHOD(AExists) {
+    static void AExists(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         A_METHOD_PROLOG(3);
 
-        NanUtf8String _path (args[0]->ToString());
-        bool watch = args[1]->ToBoolean()->BooleanValue();
+        Nan::Utf8String _path (info[0]->ToString());
+        bool watch = info[1]->ToBoolean()->BooleanValue();
 
         METHOD_EPILOG(zoo_aexists(zk->zhandle, *_path, watch, &stat_completion, cb));
     }
 
-    static NAN_METHOD(AWExists) {
+    static void AWExists(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         AW_METHOD_PROLOG(3);
-        NanUtf8String _path (args[0]->ToString());
+        Nan::Utf8String _path (info[0]->ToString());
         METHOD_EPILOG(zoo_awexists(zk->zhandle, *_path, &watcher_fn, cbw, &stat_completion, cb));
     }
 
@@ -665,34 +666,32 @@ public:
 
         LOG_DEBUG(("rc=%d, rc_string=%s, value=%s", rc, zerror(rc), value));
 
-        argv[2] = stat != 0 ? zkk->createStatObject (stat) : NanNull().As<Object>();
+        argv[2] = stat != 0 ? zkk->createStatObject (stat) : Nan::Null().As<Object>();
 
         if (value != 0) {
-            argv[3] = BufferNew(value, value_len);
+            argv[3] = BufferNew(value, value_len).ToLocalChecked();
         } else {
-            argv[3] = NanNull().As<Object>();
+            argv[3] = Nan::Null().As<Object>();
         }
 
         CALLBACK_EPILOG();
     }
 
-    static NAN_METHOD(Delete) {
-        NanScope();
-
-        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(args.This());   
+    static void Delete(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(info.This());   
         assert(zk);
-        NanUtf8String _path (args[0]->ToString());
-        uint32_t version = args[1]->ToUint32()->Uint32Value();
-  
+        Nan::Utf8String _path (info[0]->ToString());
+        uint32_t version = info[1]->ToUint32()->Uint32Value();
+
         int ret = zoo_delete(zk->zhandle, *_path, version);
-        NanReturnValue(NanNew<Int32>(ret));
+        RETURN_VALUE(info, Nan::New<Int32>(ret));
     }
 
-    static NAN_METHOD(AGet) {
+    static void AGet(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         A_METHOD_PROLOG(3);
 
-        NanUtf8String _path (args[0]->ToString());
-        bool watch = args[1]->ToBoolean()->BooleanValue();
+        Nan::Utf8String _path (info[0]->ToString());
+        bool watch = info[1]->ToBoolean()->BooleanValue();
 
         METHOD_EPILOG(zoo_aget(zk->zhandle, *_path, watch, &data_completion, cb));
     }
@@ -702,25 +701,25 @@ public:
         WATCHER_CALLBACK_EPILOG();
     }
 
-    static NAN_METHOD(AWGet) {
+    static void AWGet(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         AW_METHOD_PROLOG(3);
 
-        NanUtf8String _path (args[0]->ToString());
+        Nan::Utf8String _path (info[0]->ToString());
 
         METHOD_EPILOG(zoo_awget(zk->zhandle, *_path, &watcher_fn, cbw, &data_completion, cb));
     }
 
-    static NAN_METHOD(ASet) {
+    static void ASet(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         A_METHOD_PROLOG(4);
 
-        NanUtf8String _path (args[0]->ToString());
-        uint32_t version = args[2]->ToUint32()->Uint32Value();
+        Nan::Utf8String _path (info[0]->ToString());
+        uint32_t version = info[2]->ToUint32()->Uint32Value();
 
-        if (Buffer::HasInstance(args[1])) { // buffer
-            Local<Object> _data = args[1]->ToObject();
+        if (Buffer::HasInstance(info[1])) { // buffer
+            Local<Object> _data = info[1]->ToObject();
             METHOD_EPILOG(zoo_aset(zk->zhandle, *_path, BufferData(_data), BufferLength(_data), version, &stat_completion, cb));
         } else {    // other
-            NanUtf8String _data(args[1]->ToString());
+            Nan::Utf8String _data(info[1]->ToString());
             METHOD_EPILOG(zoo_aset(zk->zhandle, *_path, *_data, _data.length(), version, &stat_completion, cb));
         }
     }
@@ -731,31 +730,31 @@ public:
         LOG_DEBUG(("rc=%d, rc_string=%s", rc, zerror(rc)));
 
         if (strings != NULL) {
-            Local<Array> ar = NanNew<Array>((uint32_t)strings->count);
+            Local<Array> ar = Nan::New<Array>((uint32_t)strings->count);
             for (uint32_t i = 0; i < (uint32_t)strings->count; ++i) {
-                ar->Set(i, NanNew<String>(strings->data[i]));
+                ar->Set(i, LOCAL_STRING(strings->data[i]));
             }
             argv[2] = ar;
         } else {
-            argv[2] = NanNull().As<Object>();
+            argv[2] = Nan::Null().As<Object>();
         }
 
         CALLBACK_EPILOG();
     }
 
-    static NAN_METHOD(AGetChildren) {
+    static void AGetChildren(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         A_METHOD_PROLOG(3);
 
-        NanUtf8String _path (args[0]->ToString());
-        bool watch = args[1]->ToBoolean()->BooleanValue();
+        Nan::Utf8String _path (info[0]->ToString());
+        bool watch = info[1]->ToBoolean()->BooleanValue();
 
         METHOD_EPILOG(zoo_aget_children(zk->zhandle, *_path, watch, &strings_completion, cb));
     }
 
-    static NAN_METHOD(AWGetChildren) {
+    static void AWGetChildren(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         AW_METHOD_PROLOG(3);
 
-        NanUtf8String _path (args[0]->ToString());
+        Nan::Utf8String _path (info[0]->ToString());
 
         METHOD_EPILOG(zoo_awget_children(zk->zhandle, *_path, &watcher_fn, cbw, &strings_completion, cb));
     }
@@ -766,51 +765,51 @@ public:
         LOG_DEBUG(("rc=%d, rc_string=%s", rc, zerror(rc)));
 
         if (strings != NULL) {
-            Local<Array> ar = NanNew<Array>((uint32_t)strings->count);
+            Local<Array> ar = Nan::New<Array>((uint32_t)strings->count);
             for (uint32_t i = 0; i < (uint32_t)strings->count; ++i) {
-                ar->Set(i, NanNew<String>(strings->data[i]));
+                ar->Set(i, LOCAL_STRING(strings->data[i]));
             }
             argv[2] = ar;
         } else {
-            argv[2] = NanNull().As<Object>();
+            argv[2] = Nan::Null().As<Object>();
         }
 
-        argv[3] = (stat != 0 ? zkk->createStatObject (stat) : NanNull().As<Object>());
+        argv[3] = (stat != 0 ? zkk->createStatObject (stat) : Nan::Null().As<Object>());
 
         CALLBACK_EPILOG();
     }
 
-    static NAN_METHOD(AGetChildren2) {
+    static void AGetChildren2(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         A_METHOD_PROLOG(3);
 
-        NanUtf8String _path (args[0]->ToString());
-        bool watch = args[1]->ToBoolean()->BooleanValue();
+        Nan::Utf8String _path (info[0]->ToString());
+        bool watch = info[1]->ToBoolean()->BooleanValue();
 
         METHOD_EPILOG(zoo_aget_children2(zk->zhandle, *_path, watch, &strings_stat_completion, cb));
     }
 
-    static NAN_METHOD(AWGetChildren2) {
+    static void AWGetChildren2(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         AW_METHOD_PROLOG(3);
 
-        NanUtf8String _path (args[0]->ToString());
+        Nan::Utf8String _path (info[0]->ToString());
 
         METHOD_EPILOG(zoo_awget_children2(zk->zhandle, *_path, &watcher_fn, cbw, &strings_stat_completion, cb));
     }
 
-    static NAN_METHOD(AGetAcl) {
+    static void AGetAcl(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         A_METHOD_PROLOG(2);
 
-        NanUtf8String _path (args[0]->ToString());
+        Nan::Utf8String _path (info[0]->ToString());
 
         METHOD_EPILOG(zoo_aget_acl(zk->zhandle, *_path, &acl_completion, cb));
     }
 
-    static NAN_METHOD(ASetAcl) {
+    static void ASetAcl(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         A_METHOD_PROLOG(4);
 
-        NanUtf8String _path (args[0]->ToString());
-        uint32_t _version = args[1]->ToUint32()->Uint32Value();
-        Local<Array> arr = Local<Array>::Cast(args[2]);
+        Nan::Utf8String _path (info[0]->ToString());
+        uint32_t _version = info[1]->ToUint32()->Uint32Value();
+        Local<Array> arr = Local<Array>::Cast(info[2]);
 
         struct ACL_vector *aclv = zk->createAclVector(arr);
 
@@ -822,19 +821,19 @@ public:
         METHOD_EPILOG(zoo_aset_acl(zk->zhandle, *_path, _version, aclv, void_completion, data));
     }
 
-    static NAN_METHOD(ASync) {
+    static void ASync(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         A_METHOD_PROLOG(2);
 
-        NanUtf8String _path (args[0]->ToString());
+        Nan::Utf8String _path (info[0]->ToString());
 
         METHOD_EPILOG(zoo_async(zk->zhandle, *_path, &string_completion, cb));
     }
 
-    static NAN_METHOD(AddAuth) {
+    static void AddAuth(const Nan::FunctionCallbackInfo<v8::Value>& info) {
         A_METHOD_PROLOG(3);
 
-        NanUtf8String _scheme (args[0]->ToString());
-        NanUtf8String _auth (args[1]->ToString());
+        Nan::Utf8String _scheme (info[0]->ToString());
+        Nan::Utf8String _auth (info[1]->ToString());
 
         struct completion_data *data = (struct completion_data *) malloc(sizeof(struct completion_data));
         data->cb = cb;
@@ -845,26 +844,26 @@ public:
     }
 
     Local<Object> createAclObject (struct ACL_vector *aclv) {
-        NanEscapableScope();
+        Nan::EscapableHandleScope scope;
 
-        Local<Array> arr = NanNew<Array>(aclv->count);
+        Local<Array> arr = Nan::New<Array>(aclv->count);
 
         for (int i = 0; i < aclv->count; i++) {
             struct ACL *acl = &aclv->data[i];
 
-            Local<Object> obj = NanNew<Object>();
-            obj->ForceSet(NanNew<String>("perms"), NanNew<Integer>(acl->perms), ReadOnly);
-            obj->ForceSet(NanNew<String>("scheme"), NanNew<String>(acl->id.scheme), ReadOnly);
-            obj->ForceSet(NanNew<String>("auth"), NanNew<String>(acl->id.id), ReadOnly);
+            Local<Object> obj = Nan::New<Object>();
+            Nan::ForceSet(obj, LOCAL_STRING("perms"), Nan::New<Integer>(acl->perms), ReadOnly);
+            Nan::ForceSet(obj, LOCAL_STRING("scheme"), LOCAL_STRING(acl->id.scheme), ReadOnly);
+            Nan::ForceSet(obj, LOCAL_STRING("auth"), LOCAL_STRING(acl->id.id), ReadOnly);
 
             arr->Set(i, obj);
         }
 
-        return NanEscapeScope(arr);
+        return scope.Escape(arr);
     };
 
     struct ACL_vector *createAclVector (Handle<Array> arr) {
-        NanScope();
+        Nan::HandleScope scope;
 
         struct ACL_vector *aclv = (struct ACL_vector *) malloc(sizeof(struct ACL_vector));
         aclv->count = arr->Length();
@@ -873,9 +872,9 @@ public:
         for (int i = 0, l = aclv->count; i < l; i++) {
             Local<Object> obj = Local<Object>::Cast(arr->Get(i));
 
-            NanUtf8String _scheme (obj->Get(NanNew<String>("scheme"))->ToString());
-            NanUtf8String _auth (obj->Get(NanNew<String>("auth"))->ToString());
-            uint32_t _perms = obj->Get(NanNew<String>("perms"))->ToUint32()->Uint32Value();
+            Nan::Utf8String _scheme (obj->Get(LOCAL_STRING("scheme"))->ToString());
+            Nan::Utf8String _auth (obj->Get(LOCAL_STRING("auth"))->ToString());
+            uint32_t _perms = obj->Get(LOCAL_STRING("perms"))->ToUint32()->Uint32Value();
 
             struct Id id;
             struct ACL *acl = &aclv->data[i];
@@ -895,8 +894,8 @@ public:
         LOG_DEBUG(("rc=%d, rc_string=%s, acl_vector=%lp", rc, zerror(rc), acl));
         CALLBACK_PROLOG(4);
 
-        argv[2] = acl != NULL ? zkk->createAclObject(acl) : NanNull().As<Object>();
-        argv[3] = stat != NULL ? zkk->createStatObject(stat) : NanNull().As<Object>();
+        argv[2] = acl != NULL ? zkk->createAclObject(acl) : Nan::Null().As<Object>();
+        argv[3] = stat != NULL ? zkk->createStatObject(stat) : Nan::Null().As<Object>();
 
         deallocate_ACL_vector(acl);
 
@@ -904,41 +903,36 @@ public:
     }
 
     static NAN_PROPERTY_GETTER(StatePropertyGetter) {
-        NanScope();
-        assert(args.This().IsEmpty() == false);
-        assert(args.This()->IsObject());
-        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(args.This());
+        assert(info.This().IsEmpty() == false);
+        assert(info.This()->IsObject());
+        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(info.This());
         assert(zk);
-        assert(NanObjectWrapHandle(zk) == args.This());
-        NanReturnValue(NanNew<Integer> (zk->zhandle != 0 ? zoo_state(zk->zhandle) : 0));
+        assert(zk->handle() == info.This());
+        RETURN_VALUE(info, Nan::New<Integer> (zk->zhandle != 0 ? zoo_state(zk->zhandle) : 0));
     }
 
     static NAN_PROPERTY_GETTER(ClientidPropertyGetter) {
-        NanScope();
-        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(args.This());
+        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(info.This());
         assert(zk);
-        NanReturnValue(zk->idAsString(zk->zhandle != 0 ? zoo_client_id(zk->zhandle)->client_id : zk->myid.client_id));
+        RETURN_VALUE(info, zk->idAsString(zk->zhandle != 0 ? zoo_client_id(zk->zhandle)->client_id : zk->myid.client_id));
     }
 
     static NAN_PROPERTY_GETTER(ClientPasswordPropertyGetter) {
-        NanScope();
-        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(args.This());
+        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(info.This());
         assert(zk);
-        NanReturnValue(zk->PasswordToHexString(zk->zhandle != 0 ? zoo_client_id(zk->zhandle)->passwd : zk->myid.passwd));
+        RETURN_VALUE(info, zk->PasswordToHexString(zk->zhandle != 0 ? zoo_client_id(zk->zhandle)->passwd : zk->myid.passwd));
     }
 
     static NAN_PROPERTY_GETTER(SessionTimeoutPropertyGetter) {
-        NanScope();
-        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(args.This());
+        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(info.This());
         assert(zk);
-        NanReturnValue(NanNew<Integer> (zk->zhandle != 0 ? zoo_recv_timeout(zk->zhandle) : -1));
+        RETURN_VALUE(info, Nan::New<Integer> (zk->zhandle != 0 ? zoo_recv_timeout(zk->zhandle) : -1));
     }
 
     static NAN_PROPERTY_GETTER(IsUnrecoverablePropertyGetter) {
-        NanScope();
-        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(args.This());
+        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(info.This());
         assert(zk);
-        NanReturnValue(NanNew<Integer> (zk->zhandle != 0 ? is_unrecoverable(zk->zhandle) : 0));
+        RETURN_VALUE(info, Nan::New<Integer> (zk->zhandle != 0 ? is_unrecoverable(zk->zhandle) : 0));
     }
 
     void realClose (int code) {
@@ -963,17 +957,16 @@ public:
                 uv_poll_stop(&zk_io);
             }
             Unref();
-            NanScope();
-            DoEmitClose (NanNew(on_closed), code);
+            Nan::HandleScope scope;
+            DoEmitClose (Nan::New(on_closed), code);
         }
     }
 
-    static NAN_METHOD(Close) {
-        NanScope();
-        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(args.This());
+    static void Close(const Nan::FunctionCallbackInfo<v8::Value>& info) {
+        ZooKeeper *zk = ObjectWrap::Unwrap<ZooKeeper>(info.This());
         assert(zk);
         zk->realClose(0);
-        NanReturnThis();
+        RETURN_THIS(info);
     };
 
     virtual ~ZooKeeper() {
@@ -1003,19 +996,19 @@ private:
 } // namespace "zk"
 
 extern "C" void init(Handle<Object> target) {
-  INITIALIZE_STRING (zk::on_closed,            "close");
-  INITIALIZE_STRING (zk::on_connected,         "connect");
-  INITIALIZE_STRING (zk::on_connecting,        "connecting");
-  INITIALIZE_STRING (zk::on_event_created,     "created");
-  INITIALIZE_STRING (zk::on_event_deleted,     "deleted");
-  INITIALIZE_STRING (zk::on_event_changed,     "changed");
-  INITIALIZE_STRING (zk::on_event_child,       "child");
-  INITIALIZE_STRING (zk::on_event_notwatching, "notwatching");
+    INITIALIZE_STRING (zk::on_closed,            "close");
+    INITIALIZE_STRING (zk::on_connected,         "connect");
+    INITIALIZE_STRING (zk::on_connecting,        "connecting");
+    INITIALIZE_STRING (zk::on_event_created,     "created");
+    INITIALIZE_STRING (zk::on_event_deleted,     "deleted");
+    INITIALIZE_STRING (zk::on_event_changed,     "changed");
+    INITIALIZE_STRING (zk::on_event_child,       "child");
+    INITIALIZE_STRING (zk::on_event_notwatching, "notwatching");
 
-  INITIALIZE_SYMBOL (zk::HIDDEN_PROP_ZK);
-  INITIALIZE_SYMBOL (zk::HIDDEN_PROP_HANDBACK);
+    INITIALIZE_SYMBOL (zk::HIDDEN_PROP_ZK);
+    INITIALIZE_SYMBOL (zk::HIDDEN_PROP_HANDBACK);
 
-  zk::ZooKeeper::Initialize(target);
+    zk::ZooKeeper::Initialize(target);
 }
 
 NODE_MODULE(zookeeper, init)

